src/Microsoft.ML.TimeSeries/SeasonalityDetector.cs

```diff
@@ -0,0 +1,317 @@
+        /// <param name="timeSeriesLength">the original time series length</param>
+        /// <param name="bestFrequency">the frequency with highest energy</param>
+        /// <param name="secondFrequency">the frequency with second highest energy</param>
+        private static void FindBestTwoFrequencies(Complex[] values, int timeSeriesLength, out int bestFrequency, out int secondFrequency)
+        {
+            bestFrequency = -1;
+            double bestEnergy = -1.0;
+            secondFrequency = -1;
+            double secondEnergy = -1.0;
+
+            if (values.Length < 2)
+                return;
+
+            List<double> energies = new List<double>();
+
+            /* Length of time series divided by frequency is period.
+             * It is obvious that the period should be larger than 1 and smaller than the total length, and is an integer.
+             */
+            for (int i = values.Length / timeSeriesLength; i < values.Length / 2 + 1; i++)
+            {
+                double nextWeight = values[i].Magnitude;
+                energies.Add(nextWeight);
+
+                if (nextWeight > bestEnergy)
+                {
+                    bestEnergy = nextWeight;
+                    bestFrequency = i;
+                }
+            }
+
+            /* Once we found a best frequency, the region formed by lower bound to upper bound corresponding to this frequency
+             * will not be inspected anymore. because they all share the same period.
+            */
+            int period = values.Length / bestFrequency;
+            double lowerBound = values.Length * 1.0 / (period + 1);
+            double upperBound = values.Length * 1.0 / (period - 1);
+
+            for (int i = values.Length / timeSeriesLength; i < values.Length / 2 + 1; i++)
+            {
+                if (i > lowerBound && i < upperBound)
+                    continue;
+                double weight = values[i].Magnitude;
+                if (weight > secondEnergy)
+                {
+                    double prevWeight = 0;
+                    if (i > 0)
+                        prevWeight = values[i - 1].Magnitude;
+                    double nextWeight = 0;
+                    if (i < values.Length - 1)
+                        nextWeight = values[i + 1].Magnitude;
+
+                    // should be a local maximum
+                    if (weight >= prevWeight && weight >= nextWeight)
+                    {
+                        secondEnergy = nextWeight;
+                        secondFrequency = i;
+                    }
+                }
+            }
+
+            var medianAggregator = new MedianDblAggregator(energies.Count);
+            foreach (var value in energies)
+            {
+                medianAggregator.ProcessValue(value);
+            }
+
+            var typycalEnergy = medianAggregator.Median;
```

Comment:
Is it possible to combine this block with where you are adding to the energies list? If all you need is the median energy then you may be able to get away with not having the energies list and directly using the mediagAggregator

Reply:
good point, remove the energies list.